instrumentation/opentelemetry-instrumentation-aws-lambda/src/opentelemetry/instrumentation/aws_lambda/__init__.py

```diff
@@ -0,0 +1,116 @@
+    # Ref Doc: https://docs.aws.amazon.com/lambda/latest/dg/lambda-python.html
+
+    import boto3
+    from opentelemetry.instrumentation.aws_lambda import (
+        AwsLambdaInstrumentor
+    )
+
+    # Enable instrumentation
+    AwsLambdaInstrumentor().instrument()
+
+    # Lambda function
+    def lambda_handler(event, context):
+        s3 = boto3.resource('s3')
+        for bucket in s3.buckets.all():
+            print(bucket.name)
+
+        return "200 OK"
+
+API
+---
+"""
+
+import logging
+import os
+from importlib import import_module
+
+from wrapt import wrap_function_wrapper
+
+from opentelemetry.instrumentation.aws_lambda.version import __version__
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.utils import unwrap
+from opentelemetry.sdk.extension.aws.trace.propagation.aws_xray_format import (
+    TRACE_HEADER_KEY,
+    AwsXRayFormat,
+)
+from opentelemetry.trace import SpanKind, get_tracer, get_tracer_provider
+from opentelemetry.trace.propagation.textmap import DictGetter
+
+logger = logging.getLogger(__name__)
+
+
+class AwsLambdaInstrumentor(BaseInstrumentor):
+    # pylint: disable=attribute-defined-outside-init
+    def _instrument(self, **kwargs):
+        self._tracer = get_tracer(
+            __name__, __version__, kwargs.get("tracer_provider")
+        )
+
+        self._tracer_provider = get_tracer_provider()
+
+        lambda_handler = os.environ.get("_HANDLER")
+        wrapped_names = lambda_handler.rsplit(".", 1)
+        self._wrapped_module_name = wrapped_names[0]
+        self._wrapped_function_name = wrapped_names[1]
+
+        wrap_function_wrapper(
+            self._wrapped_module_name,
+            self._wrapped_function_name,
+            self._function_patch,
+        )
+
+    def _uninstrument(self, **kwargs):
+        unwrap(
+            import_module(self._wrapped_module_name),
+            self._wrapped_function_name,
+        )
+
+    def _function_patch(self, original_func, _, args, kwargs):
+        lambda_context = args[1]
+        ctx_aws_request_id = lambda_context.aws_request_id
+        orig_handler = os.environ.get(
+            "ORIG_HANDLER", os.environ.get("_HANDLER")
+        )
+        xray_trace_id = os.environ.get("_X_AMZN_TRACE_ID", "")
+
+        propagator = AwsXRayFormat()
+        parent_context = propagator.extract(
+            DictGetter(), {TRACE_HEADER_KEY: xray_trace_id}
+        )
+
+        with self._tracer.start_as_current_span(
+            orig_handler, context=parent_context, kind=SpanKind.CLIENT
+        ) as span:
+            # Refer: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/faas.md#example
+            span.set_attribute("faas.execution", ctx_aws_request_id)
+
+            result = original_func(*args, **kwargs)
+
+        # force_flush before function quit in case of Lambda freeze.
+        self._tracer_provider.force_flush()
```

Comment:
Can you explain a bit more about what Lambda freeze is? We shouldn't really have any span_processor/exporter logic in an instrumentation. What if the user has other span_processors added as part of the tracer_provider?

Reply:
I agree. This shouldn't be part of the instrumentation. This should be part of some auto-instrumentation script or decorator we can provide.